content/blog/2020-03-06-wasm/index.md

```diff
@@ -0,0 +1,171 @@
+much faster than parsing JavaScript code to an IR.
+
+Then benefits of WebAssembly from actual compilation kick in. 
+JavaScript needs to be compiled to multiple versions based on what types 
+are in use(similar to any other dynamically typed language). WebAssembly 
+code has its types encoded during offline compilation on to WebAssembly.
+Therefore, it doesn't need monitoring (in the interpreter) to figure out 
+the types, and maintain multiple versions. The JavaScript engine also 
+doesn't need to do most optimizations, except for platform and hardware 
+dependent ones, as everything else is already done in static compile time.
+
+Since WebAssembly doesn't need assumptions (such as which type certain 
+object is) during interpretting, the JavaScript engine doesn't  need to 
+bail out and reoptimize as such errors never occur. 
+
+Finally, WebAssembly also allow you to manage memory manually (it only 
+supports manual memory management as of now, but automation is to be added
+ as an option) which allows you to avoid expensive garbage collection 
+during interpretation.
+
+## Implementing WebAssembly compilers
+WebAssembly is walking a tight rope between high performance (C world), and 
+safety and portability(JS world). It is hard to theorize how to optimize for
+it, so the WebAssembly team implemented extensions in different browsers to
+validate that it is indeed possible to achieve the said goal.
+
+V8 (from Chrome) and SpiderMonkey (from Firefox), reuse their optimizing JIT 
+compilers to compile ahead-of-time. This provides them predictable high 
+performance, as opposed to the unpredictable warmup times with JavaScript.
+Chakra (from Edge) uses lazy translation and optimizes only hot code. This 
+achieves faster startup time.
+
+To permit efficient use of baseline JIT compilers, WebAssembly is desiged to
+do fast validation and ability to track registers for allocation without IR
+generation. To integrate well with optimizing JIT compilers, WebAssembly is 
+designed to do direct-to-SSA translation. Moreover, structured control flow 
+makes decoding simpler.
+
+## How good is it?
+Writing code in WebAssembly doesn't mean it'll be automatically faster. 
+JavaScript can be in theory, more performant in execution (at least for now, 
+where WebAssembly is interpretted using JavaScript enging). But this 
+requires the programmer to know JIT compilation internals and constrain to 
+one browser as each has different interpreters. WebAssembly is
+already optimized statically (more time for the compiler to optimize) for 
+the general case and each interpreter can leverage it's generality to do 
+additional optimizations better than on generic JavaScript. So in practice
+WebAssembly can be much more performant.
+
+The following bar graph demonstrates how WebAssembly performs comparative to
+native code (running a C application). Most benchmarks are within 10\% of 
+native performance. 
```

Comment:
Yeah, I was trying to motivate that too, but
- I didn’t find a good source that compares JS to native
- the paper doesn’t do it, because they compare against the state-of-the-art asm.js. I’ll try to look for a good source.